### PR TITLE
CNS-516: fix rand to ensure determinism

### PR DESCRIPTION
**Pull request by orenl-lava**
Using the global rand directly is not safe in terms of determinism, e.g. if
another go routine calls (now or in the future) it to get a random number at
the wrong time. Likewise, rand.Seed() is deprecated in Go (for very similar
reasons).
    
Also, Go does not guarantee that the PRNG driving the default source for rand
will always provide the same output for a given seed - across different Go
versions.
    
Instead, implement our own source (PRNG) that combines rand.New() to provide
a dedicated (private) instance of a PRNG that uses the xoroshiro256** PRNG
from the the xorshift package as a source.
    
(See https://en.wikipedia.org/wiki/Xorshift for more details)

### CODE REVIEW

File: x/pairing/keeper/scores/score.go [hunk 1]
```diff
@@ -189,17 +188,10 @@ func PickProviders(ctx sdk.Context, scores []*PairingScore, groupCount int, hash
...
-	for i := 0; i < len(sum256)/8; i++ {
-		seed ^= int64(binary.BigEndian.Uint64(sum256[i*8 : (i+1)*8]))
-	}
-
-	rand.Seed(seed)
+	rng := rand.New(hashData)
 
 	for it := 0; it < groupCount; it++ {
-		randomValue := uint64(rand.Int63n(scoreSum.BigInt().Int64())) + 1
+		randomValue := uint64(rng.Int63n(scoreSum.BigInt().Int64())) + 1
```
Comment (Yaroms) [inline x/pairing/keeper/scores/score.go:194]: the +1 needs to be inside the random

File: utils/rand/rand_test.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func isUniformlyDistributed(t *testing.T, f func(n int64) int64) {
+	t.Helper()
```
Comment (Yaroms) [inline utils/rand/rand_test.go:11]: just so i can learn, what does this do?

File: x/pairing/keeper/pairing_test.go [hunk 3]
```diff
@@ -777,134 +777,94 @@ func TestSelectedProvidersPairing(t *testing.T) {
...
+		mapProviders[provider.Address] = i
+	}
+
+	// calculate the total expected weight
+	var totalWeight int64
+	for _, provider := range allProviders {
+		totalWeight += weight(provider)
+	}
+
+	// calculate calculate the histogram
```
Comment (oren-lava) [inline x/pairing/keeper/pairing_test.go:800]: "calculate calculate"

**Review by oren-lava — CHANGES_REQUESTED** [review]

File: utils/rand/rand.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	"math/rand"
+
+	"github.com/vpxyz/xorshift/xoroshiro128starstar"
+)
+
+func generateSeed(data []byte) (seed int64) {
+	sum256 := sha256.Sum256(data)
+
+	// fold the SHA-256 hash into a 64-bit seed using bitwise XOR
+	for i := 0; i < len(sum256)/8; i++ {
```
Comment (Yaroms) [inline utils/rand/rand.go:15]: take only the lsb

File: x/pairing/keeper/pairing_test.go (hunk 3 above)
Comment (orenl-lava) [reply x/pairing/keeper/pairing_test.go:800]: yes, it's a lot of calculus :)

File: x/pairing/keeper/scores/score.go (hunk 1 above)
Comment (orenl-lava) [reply x/pairing/keeper/scores/score.go:194]: the +1 is at the right place as it is.
we want a random number between 1 and the scoreSum, as values 1-score-of-first-item would mean "pick the first", and the next score-of-second-item values would mean "pick the second" and so on. but `Int63n()` returns a number between 0 and scoreSum-1 - so we need to shift up by one.

File: utils/rand/rand.go (hunk 4 above)
Comment (orenl-lava) [reply utils/rand/rand.go]: fixed 21eca577

File: x/pairing/keeper/pairing_test.go (hunk 3 above)
Comment (orenl-lava) [reply x/pairing/keeper/pairing_test.go]: fixed 21eca577

**Review by Yaroms — APPROVED** [review]

**Review by Yaroms — APPROVED** [review]

**Pull request merged**